sync15-adapter/src/token.rs

```diff
@@ -20,91 +23,106 @@ header! { (XTimestamp, "X-Timestamp") => [ServerTimestamp] }
     credentials: hawk::Credentials,
+    server_timestamp: ServerTimestamp,
+    valid_until: SystemTime,
+    // A copy of the api_endpoint that's in the token - if we ever try and
+    // refresh the token and get back a different endpoint we will Err.
+    api_endpoint: String,
 }
 
 // hawk::Credentials doesn't implement debug -_-
-impl fmt::Debug for TokenserverClient {
+impl fmt::Debug for TokenserverContext {
     fn fmt(&self, f: &mut fmt::Formatter) -> ::std::result::Result<(), fmt::Error> {
-        f.debug_struct("TokenserverClient")
+        f.debug_struct("TokenserverContext")
          .field("token", &self.token)
-         .field("server_timestamp", &self.server_timestamp)
          .field("credentials", &"(omitted)")
+         .field("server_timestamp", &self.server_timestamp)
+         .field("valid_until", &self.valid_until)
          .finish()
     }
 }
 
-fn token_url(base_url: &str) -> Result<Url> {
-    let mut url = Url::parse(base_url)?;
-    // kind of gross but avoids problems if base_url has a trailing slash.
-    url.path_segments_mut()
-        // We can't do anything anyway if this is the case.
-       .map_err(|_| error::unexpected("Bad tokenserver url (cannot be base)"))?
-       .extend(&["1.0", "sync", "1.5"]);
-    Ok(url)
-}
-
-impl TokenserverClient {
-    #[inline]
-    pub fn server_timestamp(&self) -> ServerTimestamp {
-        self.server_timestamp
-    }
-
-    #[inline]
-    pub fn token(&self) -> &TokenserverToken {
-        &self.token
-    }
-
-    pub fn new(request_client: &Client, base_url: &str, access_token: String, key_id: String) -> Result<TokenserverClient> {
-        let mut resp = request_client.get(token_url(base_url)?)
-                                     .header(Authorization(Bearer { token: access_token }))
-                                     .header(XKeyID(key_id))
+impl TokenserverContext {
+    pub fn new(tsc: &TokenserverClient, request_client: &Client) -> Result<TokenserverContext> {
+        let mut resp = request_client.get(token_url(&tsc.base_url)?)
+                                     .header(Authorization(Bearer { token: tsc.access_token.clone() }))
+                                     .header(XKeyID(tsc.key_id.clone()))
                                      .send()?;
 
         if !resp.status().is_success() {
             warn!("Non-success status when fetching token: {}", resp.status());
             // TODO: the body should be JSON and contain a status parameter we might need?
             debug!("  Response body {}", resp.text().unwrap_or("???".into()));
-            if let Some(seconds) = resp.headers().get::<RetryAfter>().map(|h| **h) {
-                bail!(error::ErrorKind::BackoffError(seconds));
+            // XXX - shouldn't we "chain" these errors - ie, a BackoffError could
+            // have a TokenserverHttpError as its cause?
+            if let Some(ms) = resp.headers().get::<RetryAfter>().map(|h| (**h * 1000f64) as u64) {
```

Comment:
I think we should consider sanity checking that `**h >= 0.0`. I think we'll panic otherwise, I could be convinced this is okay but my gut says we'd rather handle it explicitly.

Admittedly this should never happen.